hail/python/test/hailtop/aiotools/test_copy.py

```diff
@@ -0,0 +1,351 @@
+                raise DidNotRaiseError()
+        elif type != self._expected_type:
+            raise RaisedWrongExceptionError(type)
+
+        # suppress exception
+        return True
+
+@pytest.mark.asyncio
+async def test_copy_doesnt_exist(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    with pytest.raises(FileNotFoundError):
+        await fs.copy(Transfer(f'{src_base}a', dest_base))
+
+
+@pytest.mark.asyncio
+async def test_copy_file(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    await create_test_file(fs, 'src', src_base, 'a')
+
+    await fs.copy(Transfer(f'{src_base}a', dest_base.rstrip('/')))
+
+    await expect_file(fs, f'{dest_base}a', 'src/a')
+
+
+@pytest.mark.asyncio
+async def test_copy_large_file(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    # mainly needs to be larger than the transfer block size (8K)
+    contents = secrets.token_bytes(1_000_000)
+    async with await fs.create(f'{src_base}a') as f:
+        await f.write(contents)
+
+    await fs.copy(Transfer(f'{src_base}a', dest_base.rstrip('/')))
+
+    async with await fs.open(f'{dest_base}a') as f:
+        copy_contents = await f.read()
+    assert copy_contents == contents
+
+
+@pytest.mark.asyncio
+async def test_copy_rename_file(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    await create_test_file(fs, 'src', src_base, 'a')
+
+    await fs.copy(Transfer(f'{src_base}a', f'{dest_base}x'))
+
+    await expect_file(fs, f'{dest_base}x', 'src/a')
+
+
+@pytest.mark.asyncio
+async def test_copy_rename_file_dest_target_file(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    await create_test_file(fs, 'src', src_base, 'a')
+
+    await fs.copy(Transfer(f'{src_base}a', f'{dest_base}x', treat_dest_as=Transfer.TARGET_FILE))
+
+    await expect_file(fs, f'{dest_base}x', 'src/a')
+
+
+@pytest.mark.asyncio
+async def test_overwrite_rename_file(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    await create_test_file(fs, 'src', src_base, 'a')
+    await create_test_file(fs, 'dest', dest_base, 'x')
+
+    await fs.copy(Transfer(f'{src_base}a', f'{dest_base}x'))
+
+    await expect_file(fs, f'{dest_base}x', 'src/a')
+
+
+@pytest.mark.asyncio
+async def test_copy_rename_dir(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    await create_test_dir(fs, 'src', src_base, 'a/')
+
+    await fs.copy(Transfer(f'{src_base}a', f'{dest_base}x'))
+
+    await expect_file(fs, f'{dest_base}x/file1', 'src/a/file1')
+    await expect_file(fs, f'{dest_base}x/subdir/file2', 'src/a/subdir/file2')
+
+
+@pytest.mark.asyncio
+async def test_copy_file_dest_trailing_slash_target_dir(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    await create_test_file(fs, 'src', src_base, 'a')
+
+    await fs.copy(Transfer(f'{src_base}a', dest_base, treat_dest_as=Transfer.TARGET_DIR))
```

Comment:
Will this error if it's `Transfer.TARGET_FILE` instead?

Reply:
I added a test_copy_dest_target_file_is_dir for this case.  It throws an IsADirectoryError currently.

We might want to change this behavior to be more consistent and aggressive.  I notice that in Copier._dest_type, I don't check the dest if it has a trailing slash or treat_dest_as == Transfer.TARGET_DIR.  But I do stat the dest if treat_dest_as == Transfer.TARGET_FILE which seems unnecessary.

If I remove this check, which has slightly higher performance, we would write the destination getting either a IsADirectoryError or creating an undetected file-and-directory situation on Google Storage.